data/account/participationRegistry.go

```diff
@@ -893,9 +966,53 @@ func (db *participationDB) GetForRound(id ParticipationID, round basics.Round) (
 
 		return nil
 	})
+	if err == ErrSecretNotFound {
+		return result, nil
+	} else if err != nil {
+		return StateProofRecordForRound{}, err
+	}
 
+	// Init stateproof fields after being able to retrieve key from database
+	result.StateProofSecrets = &merklesignature.Signer{}
+	result.StateProofSecrets.SigningKey = &crypto.FalconSigner{}
+	result.StateProofSecrets.Round = uint64(round)
+
+	err = protocol.Decode(rawStateProofKey, result.StateProofSecrets.SigningKey)
+	if err != nil {
+		return StateProofRecordForRound{}, err
+	}
+
+	var rawSignerContext []byte
+	err = db.store.Rdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		// fetch stateproof public data
+		row := tx.QueryRow(selectStateProofData, id[:])
+		err := row.Scan(&rawSignerContext)
+		if err != nil {
+			return fmt.Errorf("error while querying stateproof data: %w", err)
+		}
+		return nil
+	})
```

Comment:
Is there a reason not to keep the public keyset data in the cache along with the rest of the `Keyset` data?

Reply:
I guess it is possible and might reduce runtime of this operation (which should happen once every 256 rounds as of now), but let's do some quick calculations to see the tradeoffs:
Every `SignerContext` object contains a `merklearray.Tree`, the tree is is bound by an upper limit of 2^16 leaves, meaning around 2^17 nodes in the entire tree.
Every node is basically a `sumhash` which is defined by `SumhashDigestSize` to be 64 bytes, so we're talking `2^17 * 64 ~= 8,388,000 bytes` which is basically 8.4 megabytes per instance. Say we have a few accounts on a participating node and it adds up to be quite large.

On the other hand, reading this from the disk every ~21 minutes (I think ?) for signing a Compact Certificate might also cause some hiccup as it could be an expensive IO operation, but at least it's not happening very often.

What do you think?

@algoidan can correct me if my calculations are off